### PR TITLE
Improve `ec` traits.

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

closes: #41.

The aim is to also address #27 and #268.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: ff/src/fields/mod.rs [hunk 1]
```diff
@@ -209,6 +209,94 @@ pub trait Field:
     }
 }
 
+/// Fields that have a cyclotomic subgroup, and which can leverage efficient inversion
+/// and squaring operations for elements in this subgroup.
+pub trait CyclotomicField: Field {
```
Comment (Pratyush) [inline ff/src/fields/mod.rs:214]: New field that should be implemented for fields that have more efficient operations for their cyclotomic subgroup, such as faster inverses or squaring.

File: ff/src/fields/mod.rs [hunk 2]
```diff
@@ -209,6 +209,94 @@ pub trait Field:
...
+            let naf = crate::biginteger::arithmetic::find_wnaf(e.as_ref());
+            exp_loop(self, naf.into_iter().rev())
+        } else {
+            exp_loop(self, BitIteratorBE::without_leading_zeros(e.as_ref()).map(|e| e as i8))
+        };
+    }
+}
+
+/// Helper function to calculate the double-and-add loop for exponentiation.
+fn exp_loop<F: CyclotomicField, I: Iterator<Item = i8>>(f: &mut F, e: I) {
```
Comment (Pratyush) [inline ff/src/fields/mod.rs:274]: Helper method to abstract over the NAF and normal double-and-add loops. Can't be a closure due to weird Rust type inference issues.

File: ff/src/fields/models/fp12_2over3over2.rs [hunk 3]
```diff
@@ -50,88 +50,19 @@ impl<P: Fp12Parameters> QuadExtParameters for Fp12ParamsWrapper<P> {
     fn mul_base_field_by_frob_coeff(fe: &mut Self::BaseField, power: usize) {
         fe.mul_assign_by_fp2(Self::FROBENIUS_COEFF_C1[power % Self::DEGREE_OVER_BASE_PRIME_FIELD]);
     }
-
-    fn cyclotomic_exp(fe: &Fp12<P>, exponent: impl AsRef<[u64]>) -> Fp12<P> {
```
Comment (Pratyush) [inline ff/src/fields/models/fp12_2over3over2.rs:54]: Moved to the trait.

File: ff/src/fields/models/fp12_2over3over2.rs [hunk 4]
```diff
@@ -203,30 +134,79 @@ impl<P: Fp12Parameters> Fp12<P> {
...
+        self.c1.sub_assign(&bb);
+        self.c0 = bb;
+        self.c0 = P::mul_fp6_by_nonresidue(&self.c0);
+        self.c0.add_assign(&aa);
     }
 }
 
-// TODO: make `const fn` in 1.46.
-pub fn characteristic_square_mod_6_is_one(characteristic: &[u64]) -> bool {
+pub const fn characteristic_square_mod_6_is_one(characteristic: &[u64]) -> bool {
```
Comment (Pratyush) [inline ff/src/fields/models/fp12_2over3over2.rs:196]: This can now be a `const fn`.

File: ff/src/fields/models/fp2.rs [hunk 5]
```diff
@@ -99,3 +100,12 @@ impl<P: Fp2Parameters> Fp2<P> {
         self.c1 *= other;
     }
 }
+
+impl<P: Fp2Parameters> CyclotomicField for Fp2<P> {
+    fn cyclotomic_inverse_in_place(&mut self) -> Option<&mut Self> {
```
Comment (Pratyush) [inline ff/src/fields/models/fp2.rs:105]: We only have faster inverses here.

File: ff/src/fields/models/fp4.rs [hunk 6]
```diff
@@ -49,6 +50,15 @@ impl<P: Fp4Parameters> QuadExtParameters for Fp4ParamsWrapper<P> {
 
 pub type Fp4<P> = QuadExtField<Fp4ParamsWrapper<P>>;
 
+impl<P: Fp4Parameters> CyclotomicField for Fp4<P> {
+    fn cyclotomic_inverse_in_place(&mut self) -> Option<&mut Self> {
```
Comment (Pratyush) [inline ff/src/fields/models/fp4.rs:54]: Same for Fp4.

File: ff/src/fields/models/fp6_2over3.rs [hunk 7]
```diff
@@ -23,6 +23,15 @@ pub trait Fp6Parameters: 'static + Send + Sync {
     }
 }
 
+impl<P: Fp6Parameters> CyclotomicField for Fp6<P> {
+    fn cyclotomic_inverse_in_place(&mut self) -> Option<&mut Self> {
```
Comment (Pratyush) [inline ff/src/fields/models/fp6_2over3.rs:27]: I think technically we can also have cyclotomic squaring ala Fp12, but I didn't want to implement that in this PR.

File: ff/src/fields/models/quadratic_extension.rs [hunk 8]
```diff
@@ -89,34 +89,6 @@ pub trait QuadExtParameters: 'static + Send + Sync + Sized {
     /// A specializable method for multiplying an element of the base field by
     /// the appropriate Frobenius coefficient.
     fn mul_base_field_by_frob_coeff(fe: &mut Self::BaseField, power: usize);
-
-    /// A specializable method for exponentiating that is to be used
-    /// *only* when `fe` is known to be in the cyclotommic subgroup.
-    fn cyclotomic_exp(fe: &QuadExtField<Self>, exponent: impl AsRef<[u64]>) -> QuadExtField<Self> {
```
Comment (Pratyush) [inline ff/src/fields/models/quadratic_extension.rs:95]: There's no need for this to be in the `QuadExtParameters` trait anymore, which simplifies things slightly.

File: ff/src/fields/models/quadratic_extension.rs [hunk 9]
```diff
@@ -153,11 +125,6 @@ impl<P: QuadExtParameters> QuadExtField<P> {
         self.c1 = -self.c1;
     }
 
-    /// This is only to be used when the element is *known* to be in the cyclotomic subgroup.
-    pub fn cyclotomic_exp(&self, exponent: impl AsRef<[u64]>) -> Self {
-        P::cyclotomic_exp(self, exponent)
-    }
```
Comment (Pratyush) [inline ff/src/fields/models/quadratic_extension.rs:159]: Moved to the `CyclotomicField` trait.

File: ff/src/fields/mod.rs [hunk 10]
```diff
@@ -209,6 +209,94 @@ pub trait Field:
...
+        let mut result = *self;
+        result.cyclotomic_inverse_in_place()?;
+        Some(result)
+    }
+
+    /// Compute the inverse of [`self`]. See [`Self::INVERSE_IS_FAST`] for details.
+    /// Returns [`None`] if `self.is_zero()`, and [`Some`] otherwise.
+    fn cyclotomic_inverse_in_place(&mut self) -> Option<&mut Self> {
+        self.inverse_in_place()
+    }
```
Comment (Pratyush) [inline ff/src/fields/mod.rs:248]: The default implementation just uses slow inversion.

File: ff/src/fields/mod.rs [hunk 11]
```diff
@@ -209,6 +209,94 @@ pub trait Field:
...
+    fn cyclotomic_square(&self) -> Self {
+        let mut result = *self;
+        result.cyclotomic_square_in_place();
+        result
+    }
+
+    /// Square [`self`] in place. By default this is computed using [`Self::square_in_place`], but for
+    /// degree 12 extensions, this can be computed faster than normal squaring.
+    fn cyclotomic_square_in_place(&mut self) -> &mut Self {
+        self.square_in_place()
```
Comment (Pratyush) [inline ff/src/fields/mod.rs:233]: The default implementation just uses normal squaring.

File: ff/src/fields/mod.rs [hunk 12]
```diff
@@ -209,6 +209,94 @@ pub trait Field:
...
 
+/// Fields that have a cyclotomic subgroup, and which can leverage efficient inversion
+/// and squaring operations for elements in this subgroup.
+pub trait CyclotomicField: Field {
+
+    /// Is the inverse fast to compute? For example, in quadratic extensions, the inverse
+    /// can be computed at the cost of negating one coordinate, which is much faster than
+    /// standard inversion.
+    /// By default this is `false`, but should be set to `true` for quadratic extensions.
+    const INVERSE_IS_FAST: bool = false;
```
Comment (Pratyush) [inline ff/src/fields/mod.rs:220]: By default inversion is slow, unless we are in a quadratic extension, in which case it can be implemented as just conjugation.

File: ff/src/biginteger/arithmetic.rs [hunk 13]
```diff
@@ -52,7 +52,7 @@ pub(crate) fn mac_discard(a: u64, b: u64, c: u64, carry: &mut u64) {
     *carry = (tmp >> 64) as u64;
 }
 
-pub fn find_wnaf(num: &[u64]) -> Vec<i64> {
+pub fn find_wnaf(num: &[u64]) -> Vec<i8> {
```
Comment (Pratyush) [inline ff/src/biginteger/arithmetic.rs:55]: Drive-by improvement to memory consumption of `find_wnaf`.

File: ec/src/models/mod.rs [hunk 14]
```diff
@@ -1,12 +1,12 @@
...
-pub mod bls12;
-pub mod bn;
-pub mod bw6;
-pub mod mnt4;
-pub mod mnt6;
+// pub mod bls12;
+// pub mod bn;
+// pub mod bw6;
+// pub mod mnt4;
+// pub mod mnt6;
```
Comment (Pratyush) [inline ec/src/models/mod.rs:7]: TODO: revert this once things compile.

File: ec/src/models/short_weierstrass_jacobian.rs [hunk 15]
```diff
@@ -4,28 +4,26 @@ use ark_serialize::{
...
+    ops::{Add, AddAssign, MulAssign, Mul, Neg, Sub, SubAssign},
     vec::Vec,
+    rand::{
+        distributions::{Distribution, Standard},
+        Rng,
+    },
 };
 
 use ark_ff::{
-    bytes::{FromBytes, ToBytes},
```
Comment (Pratyush) [inline ec/src/models/short_weierstrass_jacobian.rs:14]: We no longer implement `FromBytes` and `ToBytes`, as part of a move towards #5.

File: ec/src/models/short_weierstrass_jacobian.rs [hunk 16]
```diff
@@ -41,37 +39,37 @@ use rayon::prelude::*;
...
     #[derivative(Debug = "ignore")]
     _params: PhantomData<P>,
 }
 
-impl<P: Parameters> PartialEq<GroupProjective<P>> for GroupAffine<P> {
-    fn eq(&self, other: &GroupProjective<P>) -> bool {
-        self.into_projective() == *other
+impl<P: Parameters> PartialEq<Projective<P>> for Affine<P> {
+    fn eq(&self, other: &Projective<P>) -> bool {
+        Projective::from(*self) == *other
```
Comment (Pratyush) [inline ec/src/models/short_weierstrass_jacobian.rs:52]: TODO: specialize this to take advantage of the fact that self.z == 1.

File: ec/src/models/short_weierstrass_jacobian.rs [hunk 17]
```diff
@@ -344,47 +338,29 @@ impl<P: Parameters> PartialEq for GroupProjective<P> {
...
+impl<P: Parameters> Default for Projective<P> {
     #[inline]
     fn default() -> Self {
         Self::zero()
     }
 }
 
-impl<P: Parameters> GroupProjective<P> {
+impl<P: Parameters> Projective<P> {
+    /// Construct a [`Self`] *without* validating that the coordinates form a valid point.
```
Comment (Pratyush) [inline ec/src/models/short_weierstrass_jacobian.rs:363]: TODO: address #27 and #268

File: ec/src/msm/mod.rs [hunk 18]
```diff
@@ -3,10 +3,7 @@ mod variable_base;
 pub use fixed_base::*;
 pub use variable_base::*;
 
-/// The result of this function is only approximately `ln(a)`
-/// [`Explanation of usage`]
-///
-/// [`Explanation of usage`]: https://github.com/scipr-lab/zexe/issues/79#issue-556220473
+/// The result of this function is only approximately `ln(a)`. See [here](https://github.com/arkworks-rs/snark/79#issue-556220473) for details.
```
Comment (Pratyush) [inline ec/src/msm/mod.rs:6]: drive-by fix

File: ec/src/pairing.rs [hunk 19]
```diff
@@ -0,0 +1,201 @@
+use ark_ff::{One, Zero, PrimeField, SquareRootField, CyclotomicField};
+use ark_std::{io::{Read, Write}, fmt::{Display, Formatter, Result as FmtResult}};
+use ark_serialize::SerializationError;
+use ark_std::rand::{Rng, distributions::{Distribution, Standard}};
+use crate::*;
+
+pub trait Pairing: Sized + 'static + Copy + Debug + Sync + Send + Eq {
```
Comment (Pratyush) [inline ec/src/pairing.rs:7]: Revampled `Pairing` trait.

File: ec/src/pairing.rs [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+    /// This is the scalar field of `Self::G1` and `Self::G2`.
+    type ScalarField: PrimeField + SquareRootField;
+
+    /// An element of G1.
+    type G1: CurveGroup<ScalarField = Self::ScalarField>
+        + MulAssign<Self::ScalarField>; // needed due to https://github.com/rust-lang/rust/issues/69640
```
Comment (Pratyush) [inline ec/src/pairing.rs:13]: Note that there is no longer a separate `G1Affine` and `G1Projective`.

File: ec/src/pairing.rs [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+impl<P: Pairing> Distribution<PairingOutput<P>> for Standard {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> PairingOutput<P> {
+        // Sample a random G1 element
+        let g1 = P::G1::rand(rng);
+        // Sample a random G2 element
+        let g2 = P::G2::rand(rng);
+        P::pairing(g1, g2)
```
Comment (Pratyush) [inline ec/src/pairing.rs:152]: Sampling a random GT element requires sampling random g1 and g2 elements, and then pairing them.

File: ec/src/pairing.rs [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+    type ScalarField = P::ScalarField;
+    type UniqueRepr = Self;
+
+    fn generator() -> Self::UniqueRepr {
+        // TODO: hardcode these values.
+        // Sample a random G1 element
+        let g1 = P::G1::generator();
+        // Sample a random G2 element
+        let g2 = P::G2::generator();
+        P::pairing(g1.into(), g2.into())
```
Comment (Pratyush) [inline ec/src/pairing.rs:166]: To obtain the GT generator, we just pair the G1 and G2 generators. Since this value is fixed, we can technically hardcode it, but that requires work, and is also unlikely to really be used anywhere.

File: ec/src/lib.rs [hunk 23]
```diff
@@ -32,328 +30,159 @@ use zeroize::Zeroize;
...
 
-pub mod group;
+pub mod pairing;
+pub use self::pairing::*;
 
 pub mod msm;
 
 pub mod wnaf;
 
-pub trait PairingEngine: Sized + 'static + Copy + Debug + Sync + Send + Eq + PartialEq {
```
Comment (Pratyush) [inline ec/src/lib.rs:41]: The `PairingEngine` trait is renamed to `Pairing` and the implementation now lives in the `pairing` module.

File: ec/src/pairing.rs (hunk 22 above)
Comment (burdges) [reply ec/src/pairing.rs:181]: https://indiv0.github.io/lazycell/lazycell/
https://doc.rust-lang.org/std/lazy/index.html

File: ec/src/pairing.rs (hunk 21 above)
Comment (burdges) [reply ec/src/pairing.rs:167]: I wonder if this impacts anything..

File: ff/src/fields/mod.rs [hunk 24]
```diff
@@ -209,6 +209,97 @@ pub trait Field:
...
+    }
+}
+
+/// Helper function to calculate the double-and-add loop for exponentiation.
+fn exp_loop<F: CyclotomicField, I: Iterator<Item = i8>>(f: &mut F, e: I) {
+    let self_inverse = if F::INVERSE_IS_FAST {
+        f.cyclotomic_inverse().unwrap() // must exist because self is not zero.
+    } else {
+        F::one()
+    };
```
Comment (ValarDragon) [inline ff/src/fields/mod.rs:281]:
```suggestion
    // If the inverse is fast and we're using wnaf, we compute the base points inverse.
    // Otherwise we do nothing with the variablel, so we default it to one.
    let self_inverse = if F::INVERSE_IS_FAST {
        f.cyclotomic_inverse().unwrap() // The inverse must exist because self is not zero.
    } else {
        F::one()
    };
```

File: ec/src/pairing.rs (hunk 21 above)
Comment (ValarDragon) [reply ec/src/pairing.rs:167]: Does it not suffice to sample a random GT element, and then take the final exponentiation on it?

File: ff/src/fields/mod.rs [hunk 25]
```diff
@@ -209,6 +209,97 @@ pub trait Field:
     }
 }
 
+/// Fields that have a cyclotomic subgroup, and which can leverage efficient inversion
+/// and squaring operations for elements in this subgroup.
```
Comment (ValarDragon) [inline ff/src/fields/mod.rs:213]:
```suggestion
/// Fields that have a cyclotomic subgroup, and which can leverage efficient inversion
/// and squaring operations for elements in this subgroup.
/// If a field has multiplicative order p^d - 1, the cyclotomic subgroups refer to subgroups of order φ_n(p),
/// for any n < d, where φ_n is the [nth cyclotomic polynomial](https://en.wikipedia.org/wiki/Cyclotomic_polynomial)
```

File: ff/src/fields/models/fp2.rs [hunk 26]
```diff
@@ -99,3 +100,13 @@ impl<P: Fp2Parameters> Fp2<P> {
         self.c1 *= other;
     }
 }
+
+impl<P: Fp2Parameters> CyclotomicField for Fp2<P> {
+    const INVERSE_IS_FAST: bool = true;
+    fn cyclotomic_inverse_in_place(&mut self) -> Option<&mut Self> {
```
Comment (ValarDragon) [inline ff/src/fields/models/fp2.rs:106]:
```suggestion
    fn cyclotomic_inverse_in_place(&mut self) -> Option<&mut Self> {
    /// As the multiplicative subgroup is of order p^2 - 1, the only non-trivial cyclotomic subgroup is of order p+1
    /// Therefore, for any element in the cyclotomic subgroup, we have that `x^(p+1) = 1`.
    /// Recall that `x^(p+1)` in a quadratic extension field is equal to the norm in the base field, so we have that
    /// `x * x.conjugate() = 1`. By uniqueness of inverses, for this subgroup, x.inverse() = x.conjugate()

```

**Comment by ValarDragon** [comment]
Cyclotomic field work LGTM, I suggest we move that to a standalone PR to avoid a massive PR at the end of the refactor that will be super hard to review

File: ec/src/pairing.rs (hunk 21 above)
Comment (Pratyush) [reply ec/src/pairing.rs:167]: That's a good idea, it should work. That can also provide us with a generator, right?

File: ec/src/pairing.rs (hunk 22 above)
Comment (Pratyush) [reply ec/src/pairing.rs:181]: ooh I wonder if we should use that for other APIs too now...

File: ec/src/pairing.rs (hunk 21 above)
Comment (burdges) [reply ec/src/pairing.rs:167]: It'd only provide a generators independent of pairing outputs, but most protocols need a generator e(G_1,G_2) if they need a G_t generator at all.

File: ec/src/pairing.rs (hunk 22 above)
Comment (burdges) [reply ec/src/pairing.rs:181]: Atomic laziness incurs a performance hit of like 2ns I hear, but afaik most protocols access their generator only rarely, so whatever.  It's more likely recomputation once per thread or once per exec causes slowness, especiall in WASM, so then the fastest would be pre-computed or if const generics work then 
```
const GENERATOR: Self::UniqueRepr = {
    let g1 = P::G1::generator();
    let g2 = P::G2::generator();
    P::pairing(g1.into(), g2.into())
};
```
but you could just leave a performance TODO for whoever cares about those cases.

As an aside, we've layers of tables that depend upon tables in https://github.com/w3f/rs-ec-perf/ so I used https://gitlab.com/okannen/static_init to avoid the 2ns performance hit, but it brings portability issues so likely unsuitable here.  As a further aside, rand had a long discussion around mutable thread locals in https://github.com/rust-random/rand/issues/968

File: ec/README.md [hunk 27]
```diff
@@ -6,18 +6,18 @@
...
-* [*Twisted Edwards*](https://github.com/arkworks-rs/algebra/blob/master/ec/src/models/twisted_edwards_extended.rs) curves. The `AffineCurve` in this case is in standard Twisted Edwards curve representation, whereas the `ProjectiveCurve` uses points in [Extended Twisted Edwards Coordinates](https://eprint.iacr.org/2008/522.pdf).
+* [*Short Weierstrass*](https://github.com/arkworks-rs/algebra/blob/master/ec/src/models/short_weierstrass.rs) curves. The underlying canonical representation uses standard short Weierstrass coordinates, while arithmetic relies on formulae over [Jacobian Coordinates](https://en.wikibooks.org/wiki/Cryptography/Prime_Curve/Jacobian_Coordinates).
+* [*Twisted Edwards*](https://github.com/arkworks-rs/algebra/blob/master/ec/src/models/twisted_edwards.rs) curves. The underlying canonical representation uses standard twisted Edwards coordinates, while arithmetic relies on formulae over [Extended Twisted Edwards Coordinates](https://eprint.iacr.org/2008/522.pdf).
```
Comment (ValarDragon) [inline ec/README.md:23]:
```suggestion
* [*Twisted Edwards*](https://github.com/arkworks-rs/algebra/blob/master/ec/src/models/twisted_edwards.rs) curves. The underlying canonical representation uses standard Twisted Edwards coordinates, while arithmetic relies on formulae over [Extended Twisted Edwards Coordinates](https://eprint.iacr.org/2008/522.pdf).
```

File: ec/src/pairing.rs [hunk 28]
```diff
@@ -0,0 +1,281 @@
...
+    #[must_use]
+    fn final_exponentiation(mlo: MillerLoopOutput<Self>) -> Option<PairingOutput<Self>>;
+
+    /// Computes a product of pairings.
+    #[must_use]
+    fn product_of_pairings<'a>(
+        i: impl IntoIterator<Item = &'a (Self::G1Prepared, Self::G2Prepared)>,
+    ) -> PairingOutput<Self> {
+        Self::final_exponentiation(Self::miller_loop(i)).unwrap()
+    }
```
Comment (bhgomes) [inline ec/src/pairing.rs:60]: Can we please use a pair of references instead of a reference to a pair?

```rust
impl IntoIterator<Item = (&'a Self::G1Prepared, &'a Self::G2Prepared)>
```

It's almost impossible to compose this API with others that involve calling the pairing algorithm more than once.

Comment (Pratyush) [reply ec/src/pairing.rs:60]: Yes, that's a good suggestion.
